### PR TITLE
WIP: Initial React rewrite

### PR DESCRIPTION
**Pull request by RomanGotsiy**
**Do not merge**
#327
Initial react rewrite.
When I started this I thought it will be just viewlayer rewrite but it appeared to be complete rewrite.

**This is not yet in the alpha stage** as there are a few things to implement before.

Some architecture decisions:
- OpenAPI spec is parsed into tree-structure of ready-to-use by UI components
- MobX as a state management
- styled-components for css (still considering other solutions)

Known issues before alpha:
- [x] OpenAPI 2 compatibility layer (via conversion)
- [x] simple compatibility layer with the old Redoc.init and <redoc> tag
- [x] sticky sidebar
- [x] read only properties are not filtered out from request samples
- [x] implement some options and deprecate others
- [x] Security Definitions component
- [ ] Update docs

Before release:
- [x] security info for operations
- [x] tests
- [ ] demo page with ability to try own spec
- [ ] Search
- [ ] move all constants from styles to theme
- [ ] esm modules + ts typings
- [ ] SSR support (store seriliazation/deserialization)
- [ ] Lazy loading

I could probably have missed something so feel free to clarify here.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,148 +1,112 @@
...
-    "build:all": "npm run clean && npm run build:prod && npm run build:module",
-    "changelog": "conventional-changelog -p angular -i CHANGELOG.md -s -r 1",
-    "stats": "webpack --config build/webpack.prod.js --json > stats.json"
+    "start": "webpack-dev-server --hot",
+    "start:perf": "webpack-dev-server --env.prod --env.perf",
+    "start:prod": "webpack-dev-server --env.prod",
+    "dev": "webpack-dashboard -- webpack-dev-server --hot",
+    "test": "jest",
+    "bundle": "webpack",
+    "bundle:prod": "webpack -p -env.prod",
```
Comment (shenghu) [inline package.json]: "bundle:prod": "webpack -p -env.prod" should be "bundle:prod": "webpack -p --env.prod",

Comment (RomanGotsiy) [reply package.json]: Yes, I've noticed the same issue but haven't tried to fix it yet. Just CTRL+C worked just fine for me (bundle actually is built and emitted to dist folder, but webpack process is not terminated by some strange reason).

Would be great if you could investigate this!

Comment (shenghu) [reply package.json]: It hangs there trying to connect 

`engine.io-client:polling-xhr xhr open GET: http://127.0.0.1:9838/socket.io/?EIO=3&transport=polling&t=LyeIbN8&b64=1 +1ms`

9838 is used by webpack-dashboard. But removing it from webpack.config.js/plugins, build process can complete.

Comment (RomanGotsiy) [reply package.json]: @shenghu thanks!

**Comment by shenghu** [comment]
@RomanGotsiy can I get v2.0.0-alpha.4 via some link like this, https://rebilly.github.io/ReDoc/releases/latest/redoc.min.js ?

**Comment by RomanGotsiy** [comment]
@shenghu yes, just use unpkg: https://unpkg.com/redoc@2.0.0-alpha.4/bundles/redoc.standalone.js

**Comment by macnibblet** [comment]
Nice work, can't wait for the first stable!

**Comment by grillorafael** [comment]
Hey @RomanGotsiy 

Using Redoc very heavily on my job :) Thanks for the project.

I would like to know what can I help with this port to react. Any set of tasks that you see as useful to get help? 

Maybe keep a separate `dev` branch for the react work might help contribution as well

**Comment by kavink** [comment]
@RomanGotsiy is it possible to override themes? e.g. if I want the page to use Material UI and App related color theme ?

Also if the React component is called is it directly possible to pass props to directly link to a subsection, As of now I see If I create a new url route I can directly go to a subsection , Wondering how to do it from another react component ?

**Comment by RomanGotsiy** [comment]
@grillorafael thanks for proposing your help!
Check out [CONTRIBUTING.md](https://github.com/Rebilly/ReDoc/blob/react-rewrite/.github/CONTRIBUTING.md) for getting started and open your PR's against `react-rewrite` branch for now.

Find issues with label `help-needed` and `react-rewrite` and answer there if you would like to work on them. I will add more issues with these labels soon.
Also, you can try fixing some issues reported in https://github.com/Rebilly/ReDoc/issues/384

@kavink 

> is it possible to override themes? e.g. if I want the page to use Material UI and App related color theme ?

It is work in progress. But I don't think it will be flexible enough for you to switch to Material UI.. I will investigate this possibility though (feel free to suggest your ideas here https://github.com/Rebilly/ReDoc/issues/421)

> Also if the React component is called is it directly possible to pass props to directly link to a subsection, As of now I see If I create a new url route I can directly go to a subsection , Wondering how to do it from another react component ?

nice idea! Will add this possibility

**Comment by MrBlenny** [comment]
I will (hopefully) have some spare time over the next week or so. Would lazy-loading be an appropriate feature for me to work on?

**Comment by RomanGotsiy** [comment]
@MrBlenny we were thinking about lazy loading and decided to release 2.0 without this for now. New ReDoc is more performant + we have Server-Side Rendering working.
E.g. the spec that was rendering 13-15 sec without lazy loading before now loads in 5-6 sec which is appropriate I think. (I will prepare an article with details about this new release).
So I am still not sure how I want it to be implemented... Let's discuss it once you get to it. Once you have time, DM me on twitter or email.
Thanks!

**Comment by leonardehrenfried** [comment]
Is it possible to try this out and is worth/helpful for people to do it?

Are there official builds around?

**Comment by leonardehrenfried** [comment]
Ok, thanks!

I took the liberty to build a little demo page of Redoc2: https://leonard.io/redoc-2-demo

It shows a few Swagger 2 and OpenApi3 schemas and should even be able to display files not hosted on my webserver (but haven't actually tried that bit).

I'm quite impressed with the Redoc2 progress - well done!

**Comment by grillorafael** [comment]
DId some tests here seems to be running all good. Going to do some tests using it as a react component as well which is one of my current use cases and I'll let you guys know

**Comment by grillorafael** [comment]
A few things I noticed:

1. When specifying `scrollYOffset`, it gets applied properly to the side bar but not to the content.
2. When mounting it within a  HashRouter, it conflicts with the outer app hash couter

Will post more as I see things here

**Pull request merged**